### ISSUE TITLE
Negotiation Failed on client following Windows v1803 update on server

### ISSUE THREAD

**Issue by jessetarot** [paramiko/paramiko#1222]
I run a Paramiko client to connect to a Bitvise SFTP server running on Windows (I think Windows 10).

On Friday, the Windows server pulled [Windows 10 Feature Update v1803](https://docs.microsoft.com/en-us/windows/whats-new/whats-new-windows-10-version-1803#security), and since then Paramiko has failed to Negotiate a connection.

This is strange, because I (and several other people) am able to connect using other SSH/SFTP clients, including from the same machine where paramiko fails.

I connect using:
```private_key_io = StringIO(private_key_string)
private_key = paramiko.RSAKey.from_private_key(private_key_io)
transport = paramiko.Transport((host, port))
transport.connect(username=user, pkey=private_key)  # Negotiation Failed occurs here.
client = paramiko.SFTPClient.from_transport(transport)
```
and receive an error:
```  File "/root/project/hub/connection.py", line 25, in sftp_context_manager
    transport.connect(username=user, pkey=private_key)
  File "/venv/lib/python3.6/site-packages/paramiko/transport.py", line 1170, in connect
    self.start_client()
  File "/venv/lib/python3.6/site-packages/paramiko/transport.py", line 544, in start_client
    raise SSHException('Negotiation failed.')
paramiko.ssh_exception.SSHException: Negotiation failed.
```

The paramiko log shows:
```
DEB [20180528-23:32:48.915] thr=1   paramiko.transport: starting thread (client mode): 0x734a1978
DEB [20180528-23:32:48.915] thr=1   paramiko.transport: Local version/idstring: SSH-2.0-paramiko_2.4.1
DEB [20180528-23:32:48.916] thr=1   paramiko.transport: Remote version/idstring: SSH-2.0-7.35 FlowSsh: Bitvise SSH Server (WinSSHD) 7.35
INF [20180528-23:32:48.916] thr=1   paramiko.transport: Connected (version 2.0, client 7.35)
DEB [20180528-23:32:49.029] thr=1   paramiko.transport: kex algos:['curve25519-sha256', 'curve25519-sha256@libssh.org', 'ecdh-sha2-1.3.132.0.10', 'ecdh-sha2-nistp256', 'ecdh-sha2-nistp384', 'ecdh-sha2-nistp521', 'diffie-hellman-group-exchange-sha256', 'diffie-hellman-group-exchange-sha1', 'diffie-hellman-group15-sha512', 'diffie-hellman-group14-sha256', 'diffie-hellman-group14-sha1', 'ext-info-s'] server key:['rsa-sha2-512', 'rsa-sha2-256', 'ssh-rsa', 'ecdsa-sha2-nistp384'] client encrypt:['aes256-gcm@openssh.com', 'aes128-gcm@openssh.com', 'aes256-ctr', 'aes192-ctr', 'aes128-ctr', '3des-ctr'] server encrypt:['aes256-gcm@openssh.com', 'aes128-gcm@openssh.com', 'aes256-ctr', 'aes192-ctr', 'aes128-ctr', '3des-ctr'] client mac:['hmac-sha2-256', 'hmac-sha1'] server mac:['hmac-sha2-256', 'hmac-sha1'] client compress:['zlib', 'none'] server compress:['zlib', 'none'] client lang:[''] server lang:[''] kex follows?False
DEB [20180528-23:32:49.029] thr=1   paramiko.transport: Kex agreed: ecdh-sha2-nistp256
DEB [20180528-23:32:49.029] thr=1   paramiko.transport: HostKey agreed: ecdsa-sha2-nistp384
DEB [20180528-23:32:49.029] thr=1   paramiko.transport: Cipher agreed: aes128-ctr
DEB [20180528-23:32:49.029] thr=1   paramiko.transport: MAC agreed: hmac-sha2-256
DEB [20180528-23:32:49.030] thr=1   paramiko.transport: Compression agreed: none
INF [20180528-23:32:49.241] thr=1   paramiko.transport: Disconnect (code 11): 
```

So, my understanding is that the problem occurs:
- while negotiating an SSH connection
- before I send Username & Private Key
- before attempting to create an SFTP connection
- after the client and server have agreed on Kex, HostKey algorithm, Cipher, Mac, No Compression

**Edit:** I have tested this with the same results using paramiko v2.3.1 and v2.4.1, all on Ubuntu 16.04

Any ideas why Paramiko may fail to negotiate a connection, whereas other clients (e.g. WinSCP) succeed following this Windows Update?

Thanks,

**Comment by radssh** [comment]
The Disconnect (code 11) in the Disconnect message is originating from the server side - it is kind enough to actually send an explicit disconnect message, and code 11 corresponds to SSH_DISCONNECT_BY_APPLICATION (RFC4253).

If you can check the server logs for more details on that, it should spell out the reason why the server is initiating a disconnect. If not, you might be able to squeeze more info from the client side by adding `Transport. set_hexdump(True)` prior to  the `connect()` call. That should dump all the network packets read/write to see if there's any meaningful chatter after the compression agreement being logged and the receive of the disconnect message.

**Comment by jessetarot** [comment]
Thanks @radssh 

This issue ended up being resolved by applying upgrading Bitvise to v7.42. (Although I still don't understand why other SFTP clients worked but Paramiko did not, especially given that the server and client had reached agreements for all protocols).

Thank you also for the tips regarding the error codes, and the hexdump.

Jesse
